### DISCUSSION TITLE
Identifiers

### DISCUSSION THREAD

**Discussion by sergiogragera** [nmanero/ddd-crud#33] [Q&A] [answered]
I have been working with **DELETE** and **UPDATE** operations and I find that we cannot define a route as /users/{id} because our identity within the model is, in User entity, a concatenated name and surname.

At this time we can send the object to delete as when we added it (params or body of request) and generate id from it, but we have adopted **HTTP REST** and we need to follow API nomenclature proposal.

I think that we need to identify the entities in another way to be able to execute **CRUD** from **REST** with a nomenclature like `/users/{id}` so the `{id}` must be safe to write in the url.

There are several alternatives, some known:

1. Generate **sequential** IDs.
2. Generate **UUIDs**.
3. Keep whoAmI as identifier, it can be one or many attributes of the class but it must be encoded in the url (**urlencode**) being like `/users/Perico%20Palotes`.
4. Keep whoAmI as identifier but generating a **hash** (md5, base64, sha1).

I think the following from each of these options:

1. We have to use a sequence and not repeat. This seems more like a low-level task.
2. I like it ([draft PR](https://github.com/nmanero/ddd-crud/pull/34)).
3. It shows sensitive data in the url but on the other hand it gives a lot of context to the resulting url (SEO) ([draft PR](https://github.com/nmanero/ddd-crud/pull/35)).
4. It is similar to option 2 but the identifier generation is based on some attributes, although the encoding algorithm may be less than optimal compared to UUID generator algorithm.

What is your opinion about this?

**Comment by nmanero** [comment, marked as answer]
I like the option 2 but I would differentiate the REST and the persistence layers.
I would use UUID for REST and if we start perstisting the data in the database, I would set the PK as an int and would keep the UUID in a unique column. The PK would be internal, only for making fast joins but I would not return it to the outside world.
